MSC2970: Remove pusher path requirement (#2970)
matrix-org/matrix-doc: proposals/2970-remove-pusher-path.md

being able to push general json blobs directly to devices. The removal of the
pusher path requirement would *greatly* increase flexibility here.

## Proposal

The requirements for the `url` parameter for [`POST
/_matrix/client/r0/pushers/set`](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-pushers-set)
are relaxed. In particular:

 * The path need no longer be `/_matrix/push/v1/notify`.
 * The requirement that the URL be `https` is extended to allow `http`, but
   strongly recommend `https`.

Some additional constraints are added to ensure compatibility between implementations:

 * The URL MUST contain a valid host, with an optional port. IP literal
   addresses are permitted, in accordance with [RFC3986
   3.2.2](https://tools.ietf.org/html/rfc3986#section-3.2.2).
 * The URL must NOT contain a query string or fragment (ie, it may not contain
   the characters `?` or `#`).
 * The URL must NOT include a "userinfo" section (ie, the host may not be
   preceded by a `user@` specifcication).
 * The URL must consist solely of ASCII characters. (Unicode hostnames should
   be punicode-encoded. Non-ASCII bytes in the path should be percent-encoded.)

>  * The URL may not exceed 8000 characters in length.

Reviewer: 8000 was picked fairly randomly based on https://tools.ietf.org/html/rfc7230#section-3.1.1.

infrastructure (ie, SSRF attacks), although this is mitigated since there is no
way for an attacker to read the response to such a request.

Although these are valid concerns, we note that Matrix Pushes are not
significantly different in this respect to many other "webhook" systems (for
example, Github's
[webhooks](https://docs.github.com/en/developers/webhooks-and-events/about-webhooks)
system and [AWS
SNS](https://docs.aws.amazon.com/sns/latest/dg/sns-http-https-endpoint-as-subscriber.html)),
which have no such constraint.

Rather, we propose that these attacks be mitigated by other controls such as
the following -- all of which are important anti-abuse controls regardless of
whether the URL path is restricted:

 * All outgoing HTTP requests from the Homeserver should be subject to an IP
   address blacklist, so that the server administrator can prevent access to
   internal resources.

 * We could require that the push endpoint respond to an empty `ping` request -
   and in particular that it must respond at the time that the pusher is
   configured. This would significantly reduce the scope for abuse by directing
   pushes to inappropriate endpoints.

>    However, it is considered out-of-scope for this MSC.

Reviewer: Thinking about this, I think it would be quite a beneficial addition. However, it would be a backwards-incompatible change to the pusher API (currently, there is no way to do a "test notification" other than by making up an event of some form), and generally needs a bit more time to write into an MSC than I have time for right now.

Reviewer 2: I was going to suggest something like this, though I thought about requiring *either* that the pusher URL ends with `/_matrix/push/v1/notify` *or* that it responds to `GET` requests with a predetermined response that proves it's expecting pushes.

Agreed this should be out-of-scope for this MSC.

Reviewer 3: As discussed above this could be done pretty easily by adding an `httpv2` pusher type? I guess this would require clients to know the types that a server supports though, which they might not currently.

Reviewer 4: @dkasak Another way to verify the pusher URL is how [JMAP does it](https://github.com/jmapio/jmap/blob/master/spec/jmap/push.mdown#pushsubscription) like a phone number verification code.
> when a subscription is created, the JMAP server immediately sends a PushVerification object to that URL

> NOT make any further requests to the URL until the client receives the push and updates the subscription with the correct verification code

path, whilst still imposing some constraints. For example, we might say that
the path must start with, or end with, `/_matrix/push/v1/notify`. Such an
appproach might help to mitigate SSRF attacks whilst still allowing some
flexibility for alternative implementations.

We consider that requiring that the path *start* with `/_matrix/etc` offers too
little flexibility for some implementations, and that requiring that it *end*
with `/_matrix/etc` offers little protection. For example, an attakcer might
configure a pusher with an endpoint
`https://example.com/app%3Fx%3D/_matrix/etc`, which then appears as a request
to `/app` with a query-string.

In short, we consider the benefits of such an approach are minor.

### Leave the HTTPS requirement in place

Arguably the relaxation of the requirement that the push endpoint be HTTPS is
orthogonal, but we consider this a good time to reconsider that requirement.

In short, there are various deployment strategies - in particular, where a push
gateway is known to be hosted on the same infrastructure as the homeserver -
where the requirement for HTTPS is unnecessarily onerous. In most scenarios,

> HTTPS is preferable for obvious reasons, but We don't consider it the job of
> the pusher API to enforce this.

Reviewer 3: ```suggestion
HTTPS is preferable for obvious reasons, but we don't consider it the job of
the pusher API to enforce this.
```

> # MSC2970: Remove pusher path requirement

Reviewer: we should have a "unstable endpoints" section, even if it's empty.

Recently synapse [removed the ability to set arbitrary pusher paths](https://github.com/matrix-org/synapse/pull/8865),
in order to follow the spec. During the time that synapse has allowed any pusher paths, it has shown
itself to be very useful, to the point where this fix for synapse to follow the spec has become a
real hindrance in some areas.

> With the need of push notifications without FCM or apples push system getting greater and greater, an

Reviewer 5: ```suggestion
With the need of push notifications without [FCM](https://firebase.google.com/docs/cloud-messaging) or [APNs](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html) getting greater and greater, an
```

elegant approach is to self-host push for your own devices and completely remove the need of any gateway.
This has been experimented with before, in projects such as [matrix-gotify](https://gitlab.com/Sorunome/matrix-gotify).
These setups, however, typically have a URL specific to a device. The requirement of pushing to

> `/_matrix/push/v1/notify` makes this requirement impossible.

Reviewer: Some concerns were expressed within the SCT (notably from @anoadragon453 and @erikjohnston) about whether gotify/UniversalPush could ever offer an API with the right shape for the push gateway. 

My understanding is that @Sorunome's gotify plugin (linked above) makes sure that it *does* provide that API, so I'm not really following that argument.

Reviewer: (it also seems like, modulo the requirement for a `rejected` response, you may not even need a gotify plugin to use matrix push with unified-push - that is probably a matter for a different MSC and/or changes to UP, though)

Additionally p2p push could push directly to an ipv6 address of a phone, if it is known. Depending on
implementation a custom path is not possible here, either.

In general, having a path requirement for the pushers basically forces app developers to implement
and host matrix-specific gateways, instead of just being able to push general json blobs directly to
your phone. The removal of a pusher path requirement would *greatly* increase flexibility here.

## Proposal

The requirement for pushers with a kind `http` to have the path `/_matrix/push/v1/notify` is removed.
The homeserver is expected to just push notifications in the specified format to the URL as given,
including host, path and query part. The requirement that pusher URLs MUST be https remains.

### Versioning

In the future a new, incompatible, push format may be introduced. As a client typically dictates what
format of push it wants, the new format could be indicated by setting the pusher kind to e.g. `httpv2`.
The client is then also expected to use a pusher URL which is capable of that new format. That way,
client, server and wherever the pusher URL points to all know what payload to expect.

## Potential issues

> None

Reviewer 6: There's almost certainly an issue that can be described here...

## Alternatives

Instead of allowing the pusher path to be anything, it could be introduced that the pusher path may
be prefixed with custom path fragments, but must still end in `/_matrix/push/v1/notify`. This approach
tries to solve future versioning of different payloads by setting the version within the path. However,
it would make matrix harder to integrate into existing push systems without the need for a gateway,
as a push server might not expect paths to be suffixed. The versioning via introducing a new pusher
kind seems appropriate either way, if the payload significantly changes. As such, it seems that allowing
any pusher paths is the less restrictive approach.

> ## Security considerations

Reviewer 6: In addition to @clokep's comment ([here](https://github.com/matrix-org/matrix-doc/pull/2970#issuecomment-816755056)) there's an added concern of being able to register pushers which point at random web services. This not only can point at internal infrastructure, but also a random webserver out in the wild. A handshake process may be in order.

Reviewer: Is the same as @clokep's comment, or somehow different? 

@clokep wrote:

> targeting arbitrary URLs.

@turt2live wrote 

>  being able to register pushers which point at random web services

in any case, IMHO: this is no different from any other service that provides webhooks: there is no requirement on the target path for a Github event notifications, for example.

Pivoting into internal infrastructure should be handled with IP blacklists etc (an effective IP blacklist is already required for URL previews.)

Other services that provide webhooks don't require a handshake, and I'm unconvinced that we need one. It sounds like an an annoyance for people to set up.

Concerns over traffic amplification should be handled with rate-limiting (and making sure that pushers that don't get the correct response get disabled), which is orthogonal to the URL path.

Reviewer 3: > Is the same as @clokep's comment, or somehow different?
> 
> @clokep wrote:
> 
> > targeting arbitrary URLs.

My intention from that blurb is pretty much the same as what @turt2live wrote.

> in any case, IMHO: this is no different from any other service that provides webhooks: there is no requirement on the target path for a Github event notifications, for example.

Yes, that's true. We should at least document that this does have security implications to it.

(P.S. Sorry for forgetting to use a thread!)

Reviewer 5: I agree. I think the security implications of sending less restricted HTTP requests from the homeserver are important to note but worth the minor risk.

Reviewer: I hope I've now described the security considerations in question.

  arrange to receive pushes on a specific path.

In general, having a path requirement for the pushers basically forces app
developers to implement and host matrix-specific gateways, instead of just
being able to push general json blobs directly to devices. The removal of the
pusher path requirement would *greatly* increase flexibility here.

## Proposal

The requirements for the `url` parameter for [`POST
/_matrix/client/r0/pushers/set`](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-pushers-set)
are relaxed. In particular:

 * The path need no longer be `/_matrix/push/v1/notify`.
 * The requirement that the URL be `https` is extended to allow `http`, but
   strongly recommend `https`.

Some additional constraints are added to ensure compatibility between implementations:

 * The URL MUST contain a valid host, with an optional port. IP literal
   addresses are permitted, in accordance with [RFC3986
   3.2.2](https://tools.ietf.org/html/rfc3986#section-3.2.2).
 * The URL must NOT contain a query string or fragment (ie, it may not contain
   the characters `?` or `#`).
 * The URL must NOT include a "userinfo" section (ie, the host may not be

>    preceded by a `user@` specifcication).

Reviewer 2: ```suggestion
   preceded by a `user@` specification).

 * The URL must consist solely of ASCII characters. (Unicode hostnames should

>    be punicode-encoded. Non-ASCII bytes in the path should be percent-encoded.)

Reviewer 2: ```suggestion
   be Punycode-encoded. Non-ASCII bytes in the path should be percent-encoded.)

 * The URL may not exceed 8000 characters in length.

### Versioning

One reason that the constraint on paths was introduced was to allow the future
introduction of new, incompatible, push formats - with the intention that the
homeserver pick the format to be used based on the path in the URL of the
pusher.

It is proposed instead that this could be done by changing the `kind` of
pusher: for example by setting it to `httpv2`.

## Potential issues

The removal of the path constraint potentially makes it easier to construct
HTTP magnification/DoS attacks, using the homeserver as a proxy.

It could also make it easier to construct attacks against internal
infrastructure (ie, SSRF attacks), although this is mitigated since there is no

> way for an attacker to read the response to such a request.

Reviewer 2: This may just be my non-native English speaking, but "mitigated" sounds to me just a wee bit too strong here given that blind SSRFs are very much a thing.

```suggestion
It could also make it easier to construct attacks against internal
infrastructure (ie, SSRF attacks). This is somewhat lessened by the fact that
an attacker has no way to read the response to such a request, but blind 
SSRF attacks are still possible.

   internal resources.

 * We could require that the push endpoint respond to an empty `ping` request -
   and in particular that it must respond at the time that the pusher is
   configured. This would significantly reduce the scope for abuse by directing
   pushes to inappropriate endpoints.

   However, it is considered out-of-scope for this MSC.

 * Outgoing requests must be subject to rate-limiting, to prevent individual
   users creating large volumes of requests. Additionally, there should be a
   per-user limit on the total number of pushers.

 * If a push endpoint does not give a valid response at all (for example, over a
   period of 7 days), the pusher should be disabled.

## Alternatives

### Partial constraints on URI paths

An alternative approach would be to allow more flexibility in the push endpoint
path, whilst still imposing some constraints. For example, we might say that
the path must start with, or end with, `/_matrix/push/v1/notify`. Such an
appproach might help to mitigate SSRF attacks whilst still allowing some
flexibility for alternative implementations.

We consider that requiring that the path *start* with `/_matrix/etc` offers too

> little flexibility for some implementaions, and that requiring that it *end*

Reviewer 2: ```suggestion
little flexibility for some implementations, and that requiring that it *end*
